src/data_loader.py

```diff
@@ -0,0 +1,160 @@
+
+
+
+class ImageSegmentaionDataset():
+    """Image Segmentation dataset"""
+    
+    def __init__(self, image_dict, loader_type, replacement=True, transform=None):
+        """
+        Args:
+        image_dict (dict): a dict of the following format:
+                        {"ships": {"imageId": "segmentation pixels"}， “w/ships”: ["imageId"]}
+        loader_type: (string)"classification" or "segmentation"
+        replacement: (bool) sample with/without replacement
+        transform (callable, optional): Optional transform to be applied on a sample.
+        """
+        self.image_dict = image_dict
+        self.loader_type = loader_type
+        self.replacement = replacement
+        # images_list is the name 
+        self.images_list = list(image_dict["ships"].keys()) + image_dict["w/ships"]
+        random.shuffle(self.images_list) 
+        self.transform = transform
+    
+    def __len__(self):
+        return len(self.images_list)
+    
+    def rle_decode(self, mask_rle, shape=(768, 768)):
+        """
+        mask_rle: run-length as string formated (start length)
+        shape: (height,width) of array to return 
+        Returns numpy array, 1 - mask, 0 - background
+        """
+        s = mask_rle.split()
+        starts, lengths = [np.asarray(x, dtype=int) for x in (s[0:][::2], s[1:][::2])]
+        starts -= 1
+        ends = starts + lengths
+        img = np.zeros(shape[0]*shape[1], dtype=np.uint8)
+        for lo, hi in zip(starts, ends):
+            img[lo:hi] = 1
+        return img.reshape(shape).T  # Needed to align to RLE direction
+    
+    def masks_as_image(self, in_mask_list):
+        # Take the individual ship masks and create a single mask array for all ships
+        all_masks = np.zeros((768, 768), dtype = np.int16)
+        #if isinstance(in_mask_list, list):
+        for mask in in_mask_list:
+            if isinstance(mask, str):
+                all_masks += self.rle_decode(mask)
+        return np.expand_dims(all_masks, -1)
+    
+    
+    def retrive_image_matrix(self, image_id):
+        # google platform (TODO!!!!!!!!!)
+        image_path = path.join(path.dirname(__file__), '../data/train_v2/', image_id)
+
+        # kaggle image path
+        #image_path = "/kaggle/input/airbus-ship-detection/train_v2/" + image_id
+        img = PIL.Image.open(image_path)
+        image_mat = np.array(img)
+        return image_mat
+        
+    def __getitem__(self, idx):
+        image_id = self.images_list[idx]
+        no_ship_list = image_dict["w/ships"]
+        image_mat = self.retrive_image_matrix(image_id)
+        # dataloader for classification
+        if self.loader_type == "classification":
+            if image_id in no_ship_list:
+                sample = {"image": image_mat, "class": 0}
+            else:
+                sample = {"image": image_mat, "class": 1}
+        
+        # dataloader for segmentation 
+        else: 
+            mask_list = image_dict["ships"][image_id]
```

Comment:
If I understand correctly, this code assumes that we always return a mask for the image from the list of images with ships. This is not the case.
Example of outputs:
tuples*

without ships:
(<768x768 matrix of 0-255 values>, <768x768 matrix of 0's>)

with ships:
(<768x768 matrix of 0-255 values>, <768x768 matrix of 0's and 1's>)

Reply:
See comment above about ```resize``` transform. In this case, the size of both mask and input image need to be scaled and equal

Reply:
More ideas. The resize transform will use bilinear upscale, so the resulting values might be between 0 and 1. To keep mask values 0 or 1, filter the output with the following threshold:
if pixel value >= 0.5 > make it 1
if pixel value < 0.5 make it 0

Reply:
1. This is indeed a bug and I've fixed it.
2.  mask: (1, 768, 768)       input: (3, 768, 768) 
3*.  As far as I concern, it is unlikely for the resulting values to be between 0 and 1, so I ignored it for now.